[3125] Backlink on location page returns no results after validation error

gpeng opened the pull request:
### Context

When selecting a 'location' filter (or provider filter as they are selected on the same page) there is a back link. We populate the backlink with the values from the original querystring that arrives with the filter page request. When an invalid form is submitted from that page the backlink is rebuilt with the now invalid parameters. If the backlink is clicked a variety of broken results pages are displayed depending on what combination of original and invalid filters are selected. This leads to either an incorrect filter being applied or in some cases no results at all and on '14xxx across England' link being displayed.

### Changes proposed in this pull request

Store the parameters from the last valid results page in a set of hidden fields prefixed with `prev_`. In the event of an invalid form being submitted from the location page, use these to build the back link.

Remove these parameters before the results page is rendered both from the back link and also from the provider links on the interstitial (non-javascript) provider list page.

### Guidance to review

**With JS:**

* Create a valid location filter
* Click 'Change location or choose a provider' and submit an empty provider filter form so as an error message is displayed. 
* Click the back link
* You should be returned to the original location filter

* Create a valid provider filter
* Click 'Change location or choose a provider' and submit an empty location filter form so as an error message is displayed. 
* Click the back link
* You should be returned to the original provider filter

**Without JS:**

Do those ^

also...

* Submit an incomplete provider search so as the provider list is displayed. 
* Check that the links for the provider work

In all cases the `prev_` parameters should have been removed by the time you return to the results page.

The start pages don't have a back link but will still get `prev_` parameters in their querystring if an invalid form is submitted. These should have been removed by the time the results page is reached and will have been set to `none` which is ignored.

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally
- [ ] Product review

gpeng commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[M] backlink on location page returns no results after validation error](https://trello.com/c/Xt2FB5um/3125-backlink-on-location-page-returns-no-results-after-validation-error)

danielburnley commented on app/controllers/concerns/filter_parameters.rb:
```diff
@@ -25,7 +19,33 @@ def deserialize_array_filter_param(param)
+
+  def merge_previous_parameters(all_parameters)
+    previous_parameters.each do |key, value|
+      next if key =~ /prev_/ && value == "none"
```
I find this really hard to parse, you could maybe extract at least the condition to a method that explained what it was doing to make it clearer

asmega replied:
would it be safer to use `/^prev_/` here instead of `/prev_/`?

misaka replied:
```suggestion
      next if key.start_with?("prev_") && value == "none"
```

danielburnley commented on app/controllers/concerns/filter_parameters.rb:
```diff
@@ -25,7 +19,33 @@ def deserialize_array_filter_param(param)
+  end
+
+  def previous_parameters
+    parameters.select { |key, _value| key.to_s =~ /^prev_/ }
```
Wonder if there's a nice way to refactor this regex that's in a lot of places too, maybe moving it to a contatnt like `PREVIOUS_PARAMETER_NAME_MATCHER` or something

misaka replied:
Also maybe worth using the method-for-the-job:

```suggestion
    parameters.select { |key, _value| key.to_s.start_with? "^prev_" }
```

danielburnley commented on app/views/result_filters/provider/new.html.erb:
```diff
@@ -10,7 +10,10 @@
         <li data-qa="provider_suggestion">
-          <%= govuk_link_to "#{provider.provider_name} (#{provider.provider_code})", results_path(filter_params.merge(query: provider.provider_name)), { class: "govuk-link", data: { qa: "provider_suggestion__link" } } %>
+          <%= govuk_link_to "#{provider.provider_name} (#{provider.provider_code})", 
+            results_path(remove_previous_parameters(filter_params.merge(query: provider.provider_name))), 
```
Might be worth moving this to a helper method - it's quite hard to read as it is here

defong replied:
I do agree with @danielburnley 

`remove_previous_parameters` & `filter_params` are both from FilterParameters`

misaka replied:
How about `filter_params_without_previous` to combine `remove_previous_parameters` and `filter_params`?

asmega commented on app/views/shared/_hidden_previous_fields.html.erb:
```diff
@@ -0,0 +1,7 @@
+<%= form.hidden_field("prev_l", value: params["prev_l"] || params["l"] || "none") %>
```
why do we have to assign `none`? can't we skip it if it's `nil`? it'll save you from checking for `none` later?

gpeng replied:
I think I had to do that as if they aren't set on the original request but then the invalid form causes the non-prev parameters to be present in the querystring on the next request, the partial will populate the `prev_` value from the new invalid parameters. `none` is significant as it indicates that there was no parameter in the last valid filter from the results page and stops it being set subsequently.

misaka commented on app/controllers/result_filters/provider_controller.rb:
```diff
@@ -18,7 +18,9 @@ def new
       elsif @provider_suggestions.count == 1
-        redirect_to results_path(filter_params.merge(query: @provider_suggestions.first.provider_name))
+        redirect_to results_path(
+          remove_previous_parameters(filter_params.merge(query: @provider_suggestions.first.provider_name)),
```
Could also be elligible for `filter_params_without_previous` (or similar), as mentioned below in the view.

misaka commented on app/controllers/concerns/filter_parameters.rb:
```diff
@@ -25,7 +19,33 @@ def deserialize_array_filter_param(param)
+    previous_parameters.each do |key, value|
+      next if key =~ /prev_/ && value == "none"
+
+      all_parameters[key.split("_").last] = value
```
```suggestion
      all_parameters[key.delete_prefix("prev_")] = value
```

To protect against a subtle gotcha in the future, esp with the params being refactored a little bit. Also better communicates intent to reader.

misaka commented on app/controllers/concerns/filter_parameters.rb:
```diff
@@ -25,7 +19,33 @@ def deserialize_array_filter_param(param)
   end
 
+  def remove_previous_parameters(all_parameters)
+    all_parameters.except(*previous_parameters.keys)
```
Minor nit: is it worth doing this?

```suggestion
    all_parameters.reject { |k,v| k.start_with? "prev_" }
```

It's more performant (fwiw here) and less cognitive load (reader doesn't need to think about `previous_parameters`).

misaka commented on app/controllers/result_filters/location_controller.rb:
```diff
@@ -19,20 +19,22 @@ def create
       if form_object.valid?
-        all_params = form_params.merge!(form_object.params)
-        redirect_to(next_step(all_params))
+        parameters_with_geocode_added_and_previous_removed = remove_previous_parameters(form_params.merge!(form_object.params))
```
Is the `merge!` here intentional? Does `form_params` get used in the view or something?

gpeng replied:
I don't think it's needed. I've removed it.

gpeng commented:
@danielburnley @asmega @misaka Thanks for the reviews. I've addressed all of the comments now I think if you would be so kind as to have another look.
